src/alarm_heap.cpp

```diff
@@ -0,0 +1,326 @@
+           (alarm_in_heap->get_pop_time() <= time_now_in_ms))
+    {
+      // While the next alarm in the heap is due to be sent, we:
+      //  - Pull out the alarm definition for its index and severity.
+      //  - Pass the alarm definition to the trap sender (which is responsible
+      //    for actually sending any INFORMs).
+      //  - Update the master view of the current severity for this alarm (by
+      //    updating the _all_alarms_state map).
+      //  - Finally, we then remove the alarm from the heap.
+      AlarmTableDef& alarm_table_def =
+              _alarm_table_defs->get_definition(alarm_in_heap->index(),
+                                                alarm_in_heap->severity());
+
+      if (alarm_table_def.is_valid())
+      {
+        _alarm_trap_sender->send_trap(alarm_table_def);
+        AllAlarmsStateMap::iterator alarm = _all_alarms_state.find(alarm_in_heap->index());
+
+        if (alarm != _all_alarms_state.end())
+        {
+          (*alarm).second->set_severity(alarm_table_def);
+        }
+      }
+
+      alarm_in_heap->remove_from_heap();
+      alarm_in_heap = (AlarmInHeap*)_alarm_heap.get_next_timer();
+    }
+
+    // The next alarm in the heap isn't due to pop yet. Wait until it's due.
+    if (alarm_in_heap)
+    {
+      // The next alarm in the heap isn't due to pop yet. Wait until it's due.
+      struct timespec time;
+      time.tv_sec = alarm_in_heap->get_pop_time() / 1000;
+      time.tv_nsec = (alarm_in_heap->get_pop_time() % 1000) * 1000000;
+      _cond->timedwait(&time);
+    }
+    else
+    {
+      // There's no alarms in the heap. Wait until we're signalled.
+      _cond->wait();
+    }
+  }
+
+  pthread_mutex_unlock(&_lock);
+}
+
+void AlarmHeap::update_alarm_in_heap(AlarmInfo* alarm_info,
+                                     AlarmDef::Severity severity,
+                                     uint64_t alarm_pop_time)
+{
+  _alarm_heap.insert(alarm_info->alarm_in_heap());
+  alarm_info->update_alarm_in_heap(severity, alarm_pop_time);
+}
+
+void AlarmHeap::issue_alarm(const std::string& issuer,
+                            const std::string& identifier)
+{
+  unsigned int index;
+  unsigned int severity;
+  if (sscanf(identifier.c_str(), "%u.%u", &index, &severity) != 2)
+  {
+    TRC_ERROR("malformed alarm identifier: %s", identifier.c_str());
+    return;
+  }
+
+  AlarmTableDef& alarm_table_def = _alarm_table_defs->get_definition(index,
+                                                                     severity);
+
+  if (alarm_table_def.is_valid())
+  {
+    pthread_mutex_lock(&_lock);
+
+    AllAlarmsStateMap::iterator alarm = _all_alarms_state.find(index);
+
+    if (alarm != _all_alarms_state.end())
+    {
+      AlarmDef::Severity severity_as_enum = (AlarmDef::Severity)severity;
+      AlarmSeverityChange severity_change =
+                  (*alarm).second->new_alarm_severity(severity_as_enum);
+
+      if (severity_change == AlarmSeverityChange::INCREASED_SEVERITY)
```

Comment:
Could you define a custom `operator >` for AlarmDef::Severity, so that this could be `if (severity > old_severity)`? That feels more natural to read - you don't have to understand (or have) `new_alarm_severity` or `AlarmSeverityChange`.